File: frontend/Concord/src/api/usersService.js
import { request } from './httpClient';

export async function getMe() {
  return request('Users/Me');
}

export async function updateMe({ Name, Surname, Username, AvatarUrl = null }) {
  return request('Users/Me', { method: 'PUT', body: { Name, Surname, Username, AvatarUrl } });
}

export async function getUserById(userId) {
  return request(`Users/${userId}`);
}

export async function searchUsers(query) {
  return request('Users/Search', { query: { query } });
}

export async function updateMyPreferences({ Locale, NotificationsMuted, NotificationsSoundEnabled }) {
  return request('Users/Me/Preferences', {
    method: 'PUT',
    body: { Locale, NotificationsMuted, NotificationsSoundEnabled },
  });
}

export async function updateMyPrivacy({ FriendRequestPrivacy, DirectMessagePrivacy, ActivityVisibility, ReadReceiptsEnabled }) {
  return request('Users/Me/Privacy', {
    method: 'PUT',
    body: { FriendRequestPrivacy, DirectMessagePrivacy, ActivityVisibility, ReadReceiptsEnabled },
  });
}

export async function updateMyCustomStatus({ Emoji, Text, ExpiryPreset }) {
  return request('Users/Me/CustomStatus', { method: 'PUT', body: { Emoji, Text, ExpiryPreset } });
}

File: mobile/concord/lib/api/users_service.dart
import 'api_client.dart';
import 'models/enums.dart';
import 'models/my_profile_response.dart';
import 'models/public_profile_response.dart';

class UsersService {
  UsersService(this._client);

  final ApiClient _client;

  Future<MyProfileResponse> getMe() async {
    final data = await _client.get('/Users/Me');
    return MyProfileResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<MyProfileResponse> updateMe({
    required String name,
    required String surname,
    required String username,
    String? avatarUrl,
  }) async {
    final data = await _client.put(
      '/Users/Me',
      body: {'Name': name, 'Surname': surname, 'Username': username, 'AvatarUrl': avatarUrl},
    );
    return MyProfileResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<MyProfileResponse> updatePreferences({
    String? locale,
    required bool notificationsMuted,
    required bool notificationsSoundEnabled,
  }) async {
    final data = await _client.put(
      '/Users/Me/Preferences',
      body: {
        'Locale': locale,
        'NotificationsMuted': notificationsMuted,
        'NotificationsSoundEnabled': notificationsSoundEnabled,
      },
    );
    return MyProfileResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<MyProfileResponse> updatePrivacy({
    required FriendRequestPrivacy friendRequestPrivacy,
    required DirectMessagePrivacy directMessagePrivacy,
    required ActivityVisibility activityVisibility,
    required bool readReceiptsEnabled,
  }) async {
    final data = await _client.put(
      '/Users/Me/Privacy',
      body: {
        'FriendRequestPrivacy': friendRequestPrivacy.toWire(),
        'DirectMessagePrivacy': directMessagePrivacy.toWire(),
        'ActivityVisibility': activityVisibility.toWire(),
        'ReadReceiptsEnabled': readReceiptsEnabled,
      },
    );
    return MyProfileResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<MyProfileResponse> updateCustomStatus({
    required String emoji,
    required String text,
    required CustomStatusExpiryPreset expiryPreset,
  }) async {
    final data = await _client.put(
      '/Users/Me/CustomStatus',
      body: {'Emoji': emoji, 'Text': text, 'ExpiryPreset': expiryPreset.toWire()},
    );
    return MyProfileResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<PublicProfileResponse> getUser(String userId) async {
    final data = await _client.get('/Users/$userId');
    return PublicProfileResponse.fromJson(data as Map<String, dynamic>);
  }

  Future<List<PublicProfileResponse>> search(String query) async {
    final data = await _client.get('/Users/Search', query: {'query': query});
    return (data as List<dynamic>)
        .map((e) => PublicProfileResponse.fromJson(e as Map<String, dynamic>))
        .toList();
  }
}
